Лабораторная работа 1.1, 1.2

Br0adSky opened the pull request:
Лабораторная работа 1

Br0adSky committed:
Лабораторная работа
Лабораторная работа 1

shishkin-pavel commented on data_structures_and_basic_operations/src/data_structures/words.clj:
```diff
@@ -0,0 +1,59 @@
+(ns data_structures.words (:gen-class))
+
+;;функция присоединяет букву i ко всем элементам coll
```
у def/defn есть свои докстринги, лучше такое там писать?

shishkin-pavel commented on data_structures_and_basic_operations/src/data_structures/words.clj:
```diff
@@ -0,0 +1,59 @@
+(defn create-word-by-letter
+  ([word letter]
+   (if
+     (not (= (.substring word (- (.length word) 1)) letter))
```
я в группе в телеге закрепил сообщение с немного расширенными требованиями к первой лабе =)
надо избавиться от работы со строками и работать со списками произвольных объектов
для списков `cons` приклеивает спереди, поэтому какой-то порядок полученных перестановок нам не важен

shishkin-pavel commented on data_structures_and_basic_operations/src/data_structures/words2.clj:
```diff
@@ -0,0 +1,37 @@
+    (.concat word letter)))
+
+(defn create-words-by-letter
+  ([letter words] (create-words-by-letter letter words (list)))
```
такой подход провоцирует вытаскивать ненужные (снаружи) сигнатуры в пространство имён, давай заменим на `loop`
кроме того вопрос: как `recur` находит "куда передать управление"? если бы лупа не было, какими _базовыми_ средствами мы можем его сэмулировать?

shishkin-pavel commented on data_structures_and_basic_operations/src/data_structures/words2.clj:
```diff
@@ -0,0 +1,37 @@
+(defn create-words-by-letter
+  ([letter words] (create-words-by-letter letter words (list)))
+  ([letter words acc]
+   (if (> (count words) 0)
```
лучше всегда пользоваться наиболее слабыми но достаточными предикатами, тут `empty?` хватило бы
почему я настаиваю: конкретно в этом случае бесконечная последовательность -- заведомо некооректный инпут, но в других случаях они очень даже могут встретиться, а привычка останется

shishkin-pavel commented on data_structures_and_basic_operations/src/data_structures/words.clj:
```diff
@@ -0,0 +1,59 @@
+  (if
+    (> (count words) 0)
+    (if
+      (not (= (create-word-by-letter (first words) letter) nil))
```
сейчас получается, что мы сначала пытаемся сделать новое слово, а потом проверяем, получилось ли
на мой взгляд, было бы гораздо лучше сначала понять, надо ли получать слово в текущем состоянии, а не проверять на nil (кстати, почему именно нил?)

shishkin-pavel reviewed (changes requested):
с одной стороны я понимаю, что оно сделано и работает, но всё-таки хочу, чтобы ты всё поменял :)
если посмотреть на первую лабу целиком, то видно постепенное улучшение: просто рекурсия - хвостовая - высокоуровневые операции (map/filter/...)
твоё решение очень плохо выражается в терминах мэпов и других обходов коллекций
например, оно оперирует индексами, несколько ифов не имеют else-ветки, из-за чего функции, по сути, становятся не тотальными (а это опять же плохо для редьюсов)

я бы предложил опираться на такие функции:
* из слова и алфавита получаем коллекцию всех подходящих новых слов
* из набора слов и алфавила получаем коллекцию новых слов (за счёт первой функции)
* прокручиваем вторую функцию n раз: при n = 0 получаем слова длины 0, n = 1 -- слова длины 1 и так далее

если присмотреться, то это в чистом виде решение 1.4, развёрнутое из редьюсеров в примитивные функции

Br0adSky committed:
Лабораторная работа 1
Переделал согласно новым требованиям, пока черновой вариант -- надо переделать 1.2 и тесты

Br0adSky committed:
Лабораторная работа 1
Правка замечания

Br0adSky committed:
Лабораторная работа 1
Переделал 1.2 и файл утилс

Br0adSky committed:
Лабораторная работа 1
Реализация ЛР 1.3, 1.4

shishkin-pavel commented on data_structures_and_basic_operations/src/data_structures/words.clj:
```diff
@@ -0,0 +1,35 @@
+(ns data_structures.words
+  (:gen-class))
+
+(def alphabet '(:a :b :c))
```
глобальные константы это как-то фу, давай лучше всегда по вызовам протягивать
мотивация: мы же тут библиотеку для выполнения какой-то задачи делаем, зачем в ней вшиты какие-то стандартные входные данные?

shishkin-pavel commented on data_structures_and_basic_operations/src/data_structures/words.clj:
```diff
@@ -0,0 +1,35 @@
+(defn add-alphabet-to-word
+  "Добавляет алфавит к слову"
+  ([word] (add-alphabet-to-word word alphabet (list)))
+  ([word alphabet1 result]
```
тут фактически мы получаем хвостовую рекурсию (только с тратой на стек вызовов), хотя требовалось сделать через обычную (но пофиг, переделывать это не нужно)

shishkin-pavel commented on data_structures_and_basic_operations/src/data_structures/words2.clj:
```diff
@@ -0,0 +1,36 @@
+  "Проверка соотвествия длины слова"
+  [n]
+  (loop [words (add-alphabet-to-word (list))]
+   (if (= (count (first words)) n)
```
из-за того, что кложуровские списки -- линкедлисты, тут мы будем бегать по n локациям в памяти
возможно, более удачным решением было бы просто декрементировать n и сравнивать его с нулём (т.к. по построению мы гарантируем, что каждый вызов `increment-words` увеличивает длину слов на единицу)

shishkin-pavel reviewed (changes requested):
тут немного правок и давай остатки (1.3, 1.4) тоже в это ревью, а второе закроем (в нём ничего сейчас писать не буду)

Br0adSky committed:
Лабораторная работа 1
Правки МР

shishkin-pavel commented on data_structures_and_basic_operations/src/basic_operations/map_filter.clj:
```diff
@@ -0,0 +1,21 @@
+
+(defn iterate-words
+  "Увеличивает число слов, пока длина не станет = n"
+  [n] (nth (iterate increment-words (add-new-word alphabet (list))) (dec n)))
```
не потому ли потребовалось декрементировать `n`, что одну итерацию общего алгоритма (`increment-words`) ты сделал вручную за счёт `(add-new-word alphabet (list))`? давай попробуем избавиться от этого вызова и декремента, ведь у нас есть хорошее стартовое значение
кроме того, такой код не должен сработать для n = 0, и это всё очень тесно связано :)
в итоге должно получиться такое:
`(nth (iterate increment-words <X>) n)`
где `X` -- какое-то тривиальное значение, внезапно являющееся корректным ответом для n = 0

shishkin-pavel commented on data_structures_and_basic_operations/src/basic_operations/my_map_filter.clj:
```diff
@@ -0,0 +1,15 @@
+(ns basic-operations.my_map_filter)
+
+(defn my-map [function coll]
```
про 1.3 писал в чате, конкат норм, просто нужно понимание, почему оно не будет квадратичным (переписывать не обязательно, но ответ нужен)

Br0adSky replied:
ответ на что?
в чате вроде не было вопросов

Br0adSky replied:
Вопрос в том, почему с конкатом в кложе более-менее норм? Я, когда разбирался, читал, что он возвращает ленивую последовательность, также как мап и фильтр, но не могу придумать пока, как это решает проблему связанную с O(N^2). Разве что мы можем частичные вычисления запросить какие-нибудь...

shishkin-pavel replied:
да, именно про это
возьмём за основу `(map inc `(1 2 3 4))`
с обычной последовательностью нам необходимо составлять такие последовательности и каждый раз это создание всех пар коллекции заново, то мы создаём вот так
* (2)          -- +1 пара
* (2 3)       -- +2
* (2 3 4)    -- +3
* (2 3 4 5) -- +4
то в случае с ленивыми последовательностями мы можем конец "приписывать" каждый раз за O(1) за счёт отложенных вычислений (сильно уверен, что в этом месте там какая-нибудь мутабельная ячейка на конце, которая вычисляется по запросу в очередной элемент и новую мутабельную ячейку-конец)

Br0adSky committed:
Лабораторная работа 1
Правки МР

Br0adSky committed:
Лабораторная работа 1
Протащил алфавит
